Add dedicated e2e tester cmd for parallel conformance testing of kubermatic clusters

mrIncompetent opened the pull request:
**What this PR does / why we need it**:
Adds a dedicated command to run multiple e2e tests in parallel.
It also adds retries in case a particular test fails.
There are 2 kinds of retries:
- Single test during conformance test suite fails -> retries are being done by Ginkgo
- Single test fails even after Ginkgo retries: command will restart the whole suite

Test results are being stored into a dedicated directory. Test result include:
- junit.xml
- logs of conformance tests

**Release note**:
```release-note
NONE
```

mrIncompetent committed:
introduce e2e tester

mrIncompetent committed:
vendor: add ginkgo

mrIncompetent committed:
fix logging

alvaroaleman commented on api/cmd/conformance-tests/main.go:
```diff
@@ -0,0 +1,194 @@
+	scenarios = append(scenarios, getOpenStackScenarios()...)
+
+	runner := newRunner(scenarios, &opts)
+	hadFailure, err := runner.Run()
```
Why not return an error in case of failure?

alvaroaleman commented:
Awesome pull, thanks for all the work!

One nit I would like to see changed longterm is that we don't have the providers specify the scenarios themselves but instead have them global and let the providers just specify why they don't run, so its opt-out rather than opt-in.

mrIncompetent committed:
fix review feedback & add kubernetes 1.12

p0lyn0mial commented on api/cmd/conformance-tests/main.go:
```diff
@@ -0,0 +1,190 @@
+
+	if opts.cleanupOnStart {
+		if opts.namePrefix == "" {
+			log.Fatalf("cleanup-on-start was specified but name-prefix is empty")
```
replace `log.Fatal` with `glog.Fatal`

p0lyn0mial commented on api/cmd/conformance-tests/main.go:
```diff
@@ -0,0 +1,190 @@
+			}
+		}
+
+		fmt.Println("Cleaned up all old clusters")
```
`glog.Info` ?

p0lyn0mial reviewed (commented):
Nice pull indeed, brings a huge value !

- Have you considered to write the tests as standard golang tests and use `go test` to run them ? (just asking)
- Do you have a test result I could see ?

mrIncompetent commented:
@p0lyn0mial 

> Have you considered to write the tests as standard golang tests and use go test to run them ? (just asking)

Yes, i started using go tests first, but using a dedicated command felt simpler (code & usage) when using multiple flags. Thus i moved back to a simple command.

> Do you have a test result I could see ?

A test result from a scenario: https://storage.cloud.google.com/e2e-test-results/aws-ubuntu-1.10.8.zip

The folder structure is:
- reports (What you specify via flag)
  - scenario-name (For example: `aws-ubuntu-1.10.8`)
    - attempt-1
    - attempt-2 (In case a restart happened due to failures)
  - scenario-name (For example: `aws-ubuntu-1.11.3`)
    - attempt-1

mrIncompetent committed:
fix logging

mrIncompetent committed:
fix tag

mrIncompetent commented:
@alvaroaleman @p0lyn0mial 
I fixed the issues, addresses the comments and rebased into master. 
PTAL

Adding more providers & making the tests in a opt-out instead opt-int (https://github.com/kubermatic/kubermatic/pull/2112#issuecomment-426174750) will be done in follow up PR's.
